File: src/main/java/com/demonwav/mcdev/platform/mcp/at/AtUsageInspection.java
package com.demonwav.mcdev.platform.mcp.at;

import com.demonwav.mcdev.platform.MinecraftModule;
import com.demonwav.mcdev.platform.mcp.McpModule;
import com.demonwav.mcdev.platform.mcp.McpModuleType;
import com.demonwav.mcdev.platform.mcp.at.gen.psi.AtEntry;
import com.demonwav.mcdev.platform.mcp.srg.SrgManager;
import com.demonwav.mcdev.platform.mcp.srg.SrgMap;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleUtilCore;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.util.Consumer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class AtUsageInspection extends LocalInspectionTool {

    @Nullable
    @Override
    public String getStaticDescription() {
        return "The declared access transformer is never used";
    }

    @NotNull
    @Override
    public PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
        return new PsiElementVisitor() {
            @Override
            public void visitElement(PsiElement element) {
                if (!(element instanceof AtEntry)) {
                    return;
                }

                final Module module = ModuleUtilCore.findModuleForPsiElement(element);
                if (module == null) {
                    return;
                }

                final MinecraftModule instance = MinecraftModule.getInstance(module);
                if (instance == null) {
                    return;
                }

                final McpModule mcpModule = instance.getModuleOfType(McpModuleType.getInstance());
                if (mcpModule == null) {
                    return;
                }

                AtEntry entry = (AtEntry) element;

                PsiClass psiClass = entry.getClassName().getClassNameValue();
                if (psiClass == null) {
                    // Ignore here, we'll flag this as an error in another inspection
                    return;
                }

                SrgManager.getInstance(mcpModule).recomputeIfNullAndGetSrgMap().done(srgMap -> {
                    // TODO
                });
            }
        };
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/mcp/at/AtUsageInspection.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2020 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mcp.at

import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.platform.mcp.McpModuleType
import com.demonwav.mcdev.platform.mcp.at.gen.psi.AtEntry
import com.demonwav.mcdev.platform.mcp.at.gen.psi.AtFieldName
import com.demonwav.mcdev.platform.mcp.at.gen.psi.AtFunction
import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.module.ModuleUtilCore
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.search.searches.ReferencesSearch

class AtUsageInspection : LocalInspectionTool() {

    override fun getStaticDescription(): String? {
        return "The declared access transformer is never used"
    }

    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        return object : PsiElementVisitor() {
            override fun visitElement(element: PsiElement) {
                if (element !is AtEntry) {
                    return
                }

                val module = ModuleUtilCore.findModuleForPsiElement(element) ?: return
                val instance = MinecraftFacet.getInstance(module) ?: return
                val mcpModule = instance.getModuleOfType(McpModuleType) ?: return
                val srgMap = mcpModule.srgManager?.srgMapNow ?: return

                val member = element.function ?: element.fieldName ?: return
                val reference = AtMemberReference.get(element, member) ?: return

                val psi = when (member) {
                    is AtFunction ->
                        reference.resolveMember(element.project) ?: srgMap.getMcpMethod(reference)?.resolveMember(
                            element.project
                        ) ?: return
                    is AtFieldName ->
                        reference.resolveMember(element.project)
                            ?: srgMap.getMcpField(reference)?.resolveMember(element.project) ?: return
                    else ->
                        return
                }

                val query = ReferencesSearch.search(psi, GlobalSearchScope.projectScope(element.project))
                query.findFirst()
                    ?: holder.registerProblem(
                        element,
                        "Access Transformer entry is never used",
                        ProblemHighlightType.LIKE_UNUSED_SYMBOL
                    )
            }
        }
    }
}
